### PR TITLE
RTEMIS-181: Added logic for lottery process.

### PR DESCRIPTION
**Pull request by jaswinsingh**
Ticket # : https://innoraft.atlassian.net/browse/RTEMIS-181

---
### What the PR does
`(Explain fixes, changes and impacts)`
- Added logic for the lottery process.

### CODE REVIEW

File: modules/rte_mis_lottery/src/Controller/LotteryController.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      $status_code = Response::HTTP_OK;
+    }
+    else {
+      $data = [
+        '#message' => 'Access Denied',
+      ];
+      $status_code = Response::HTTP_FORBIDDEN;
+    }
+    $response = new JsonResponse($data, $status_code);
+    return $response;
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Controller/LotteryController.php:72]: Can't we directly return the jsonresponse object?

File: modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php [hunk 2]
```diff
@@ -313,26 +346,26 @@ private function createMiniNode($type = ''): void {
           'field_student_application_number' => $this->generateRandomString(11),
           'field_location' => $this->generateRandomSelectValue('field_location'),
           'field_date_of_birth' => date('2019-01-01'),
-          'field_gender' => ['transgender'],
+          'field_gender' => ['transgender', 'boy', 'girl'],
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php:349]: Can we fetch this from the configuration itself?

File: modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php [hunk 3]
```diff
@@ -392,7 +425,7 @@ private function generateParagraph($paragraph_type = '', $data = []) {
       case 'field_entry_class':
         $paragraph = Paragraph::create([
           'type' => 'entry_class',
-          'field_education_type' => array_rand(['girls', 'boys', 'co-ed']),
+          'field_education_type' => ['co-ed'],
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php:428]: Can we fetch this from configuration?

File: modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php [hunk 4]
```diff
@@ -402,19 +435,20 @@ private function generateParagraph($paragraph_type = '', $data = []) {
         break;
 
       case 'field_school_preferences':
+        // $medium_options = ['hindi', 'english'];
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php:438]: Please remove the commented code.

File: modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php [hunk 5]
```diff
@@ -402,19 +435,20 @@ private function generateParagraph($paragraph_type = '', $data = []) {
...
         $values = [
           'type' => 'school_preference',
           'field_school_id' => [
             'target_id' => $data['school_id'] ?? 1,
           ],
           'field_entry_class' => [3],
-          'field_medium' => array_rand(['hindi', 'english']),
+          // 'field_medium' => $medium_options[array_rand($medium_options)],
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php:445]: Please remove the commented code.

File: modules/rte_mis_lottery/src/Services/RteLotteryHelper.php [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    $query->fields('r', ['allotted_school_id']);
+    $query->addExpression('GROUP_CONCAT(student_id)', 'student_ids');
+    $query->groupBy('r.allotted_school_id');
+    $result = $query->execute()->fetchAll();
+    return $result;
+  }
+
+  /**
+   * Create `allotted_students_details` paragraph provided by data.
+   */
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Services/RteLotteryHelper.php:145]: Function param doc comment missing.

File: modules/rte_mis_lottery/src/Services/RteLotteryHelper.php [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+      return NULL;
+    }
+    catch (\Exception $e) {
+      return FALSE;
+    }
+  }
+
+  /**
+   * Store the lottery result for student based on allotment status.
+   */
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Services/RteLotteryHelper.php:115]: Function param doc comment missing.

File: modules/rte_mis_lottery/src/Services/RteLotteryHelper.php [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+      return FALSE;
+    }
+    catch (\Exception $e) {
+      return FALSE;
+    }
+  }
+
+  /**
+   * Update the school seat count.
+   */
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Services/RteLotteryHelper.php:88]: Function param doc comment missing.

File: modules/rte_mis_lottery/src/Services/RteLotteryHelper.php [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    catch (\Exception $e) {
+      return FALSE;
+    }
+
+  }
+
+  /**
+   * Get the current seat count for school by providing on id and entry_class.
+   */
```
Comment (kunal-sin) [inline modules/rte_mis_lottery/src/Services/RteLotteryHelper.php:66]: Function param doc comment missing.

Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Services/RteLotteryHelper.php]: Added.

File: modules/rte_mis_lottery/src/Services/RteLotteryHelper.php (hunk 8 above)
Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Services/RteLotteryHelper.php]: Added.

File: modules/rte_mis_lottery/src/Services/RteLotteryHelper.php (hunk 7 above)
Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Services/RteLotteryHelper.php]: Added.

File: modules/rte_mis_lottery/src/Controller/LotteryController.php (hunk 1 above)
Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Controller/LotteryController.php]: Done.

File: modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php (hunk 2 above)
Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php:349]: This file is only for test purpose. We can ignore this for now.

File: modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php (hunk 4 above)
Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php:438]: This file is only for test purpose. We can ignore this for now.

File: modules/rte_mis_lottery/src/Services/RteLotteryHelper.php (hunk 8 above)
Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Services/RteLotteryHelper.php]: Added

File: modules/rte_mis_lottery/src/Services/RteLotteryHelper.php (hunk 6 above)
Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Services/RteLotteryHelper.php]: Added.

File: modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php (hunk 5 above)
Comment (jaswinsingh) [reply modules/rte_mis_lottery/src/Plugin/DevelGenerate/MiniNodeDevelGenerates.php:445]: This file is only for test purpose. We can ignore this for now.

**Review by kunal-sin — APPROVED** [review]

**Pull request merged**